File: sdk/go/web/web.go
// Package web implements the Executa v2 reverse JSON-RPC
// `web/search` and `web/fetch` requests that let a plugin ask its host
// (Anna) to search the web / extract page content on its behalf.
//
// Why reverse RPC?
//   - Plugins do NOT need their own search-provider API key.
//   - Provider routing, SSRF guarding, quota and billing live in the
//     host (web_grant on UserExecuta.custom_config).
//   - Response schemas are provider-agnostic; `search_depth` expresses
//     quality intent, never provider choice.
//
// stdio-channel budgets (tighter than the iframe HOST API):
//   - fetch: max_chars ≤ 8,000/page, whole response ≤ 256 KB.
//
// Wire layout (Plugin → Agent → Nexus REST):
//
//	Plugin (us)                          Agent (host)              Nexus
//	────────────────────────────────────────────────────────────────────
//	← invoke(req_id=42, …)
//	→ web/search(req_id=A, …)            POST /copilot/web/search
//	                                      ← 200 {results, provider_tier, …}
//	← result | error
//	→ invoke result(req_id=42)
//
// Threading model identical to the sampling/image clients. Construct one
// *Client per process; feed every parsed frame to DispatchResponse.
//
// See matrix-nexus docs/design/app-web-search.md §2 / §4 for the wire
// contract.
package web

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"sync"
	"time"
)

// Constants — keep in sync with matrix/src/executa/protocol.py.
const (
	MethodWebSearch = "web/search"
	MethodWebFetch  = "web/fetch"
	// Phase 2 (app-web-search.md §2.3 / §2.4)
	MethodWebImageSearch = "web/image_search"
	MethodWebImageFetch  = "web/image_fetch"

	ErrCodeNotGranted     = -32521
	ErrCodeQuotaExceeded  = -32522 // CU pool exhausted
	ErrCodeProviderError  = -32523
	ErrCodeInvalidRequest = -32524
	ErrCodeTimeout        = -32525
	ErrCodeNotNegotiated  = -32526
	ErrCodeUserDenied     = -32527
	ErrCodeTokenExpired   = -32528
)

// WebError wraps a JSON-RPC error returned by the host.
type WebError struct {
	Code    int            `json:"code"`
	Message string         `json:"message"`
	Data    map[string]any `json:"data,omitempty"`
}

func (e *WebError) Error() string {
	return fmt.Sprintf("[%d] %s", e.Code, e.Message)
}

// SearchRequest mirrors `web/search` params (§2.1).
type SearchRequest struct {
	Query          string   `json:"query"`
	MaxResults     int      `json:"max_results,omitempty"`
	SearchDepth    string   `json:"search_depth,omitempty"` // "basic" | "advanced"
	Topic          string   `json:"topic,omitempty"`        // "general" | "news"
	TimeRange      string   `json:"time_range,omitempty"`   // day|week|month|year
	Region         string   `json:"region,omitempty"`
	IncludeDomains []string `json:"include_domains,omitempty"`
	ExcludeDomains []string `json:"exclude_domains,omitempty"`
}

// SearchResult is one item in the search results array.
type SearchResult struct {
	Title       string   `json:"title"`
	URL         string   `json:"url"`
	Snippet     string   `json:"snippet"`
	Site        string   `json:"site"`
	PublishedAt *string  `json:"published_at,omitempty"`
	Score       *float64 `json:"score,omitempty"`
}

// SearchResponse is the parsed response from `web/search`.
type SearchResponse struct {
	Results       []SearchResult `json:"results"`
	ProviderTier  string         `json:"provider_tier"` // actually-executed tier
	QuotaConsumed float64        `json:"quota_consumed"`
}

// FetchRequest mirrors `web/fetch` params (§2.2).
type FetchRequest struct {
	URLs      []string `json:"urls"`
	Format    string   `json:"format,omitempty"` // "markdown" | "text"
	MaxChars  int      `json:"max_chars,omitempty"`
	TimeoutMs int      `json:"timeout_ms,omitempty"`
}

// Page is one item in the fetch pages array (per-item failure isolation —
// check OK per page; the call itself only errors on auth/quota issues).
type Page struct {
	URL         string  `json:"url"`
	FinalURL    *string `json:"final_url,omitempty"`
	OK          bool    `json:"ok"`
	Title       *string `json:"title,omitempty"`
	Content     *string `json:"content,omitempty"`
	PublishedAt *string `json:"published_at,omitempty"`
	Truncated   bool    `json:"truncated,omitempty"`
	Error       *string `json:"error,omitempty"`
}

// FetchResponse is the parsed response from `web/fetch`.
type FetchResponse struct {
	Pages         []Page  `json:"pages"`
	QuotaConsumed float64 `json:"quota_consumed"`
}

// ImageSearchRequest mirrors `web/image_search` params (§2.3, Phase 2).
// Safe-search is always enforced host-side and cannot be disabled.
type ImageSearchRequest struct {
	Query      string `json:"query"`
	MaxResults int    `json:"max_results,omitempty"` // default 8, cap 20
	MinWidth   int    `json:"min_width,omitempty"`
	MinHeight  int    `json:"min_height,omitempty"`
	Aspect     string `json:"aspect,omitempty"` // any|wide|tall|square
}

// ImageResult is one item in the image search results array.
// LicenseHint is best-effort — the platform makes no copyright warranty.
type ImageResult struct {
	ImageURL     string  `json:"image_url"`
	ThumbnailURL *string `json:"thumbnail_url,omitempty"`
	SourceURL    string  `json:"source_url"`
	Title        *string `json:"title,omitempty"`
	Width        *int    `json:"width,omitempty"`
	Height       *int    `json:"height,omitempty"`
	MimeType     *string `json:"mime_type,omitempty"`
	LicenseHint  *string `json:"license_hint,omitempty"`
}

// ImageSearchResponse is the parsed response from `web/image_search`.
type ImageSearchResponse struct {
	Results       []ImageResult `json:"results"`
	QuotaConsumed float64       `json:"quota_consumed"`
	Cached        bool          `json:"cached,omitempty"`
}

// ImageFetchRequest mirrors `web/image_fetch` params (§2.4, Phase 2).
type ImageFetchRequest struct {
	URL      string `json:"url"`
	MaxBytes int    `json:"max_bytes,omitempty"` // default 5 MiB, cap 20 MiB
	Purpose  string `json:"purpose,omitempty"`   // audit label
}

// ImageFetchResponse is an artifact REFERENCE (never bytes — stdio safe).
// The stored object counts against the tool's storage quota.
type ImageFetchResponse struct {
	Path          string  `json:"path"`
	GetURL        string  `json:"get_url"`
	MimeType      string  `json:"mime_type"`
	BytesSize     int64   `json:"bytes_size"`
	SHA256        string  `json:"sha256"`
	SourceURL     string  `json:"source_url"`
	FinalURL      string  `json:"final_url"`
	QuotaConsumed float64 `json:"quota_consumed"`
}

// FrameWriter writes one newline-delimited JSON-RPC frame to the host.
type FrameWriter func(msg map[string]any) error

// DefaultFrameWriter writes to os.Stdout under a process-wide mutex.
func DefaultFrameWriter() FrameWriter {
	var mu sync.Mutex
	return func(msg map[string]any) error {
		buf, err := json.Marshal(msg)
		if err != nil {
			return err
		}
		mu.Lock()
		defer mu.Unlock()
		if _, err := os.Stdout.Write(append(buf, '\n')); err != nil {
			return err
		}
		return nil
	}
}

type pending struct {
	ch chan json.RawMessage
}

// Client tracks outstanding reverse RPC requests and resolves them as
// responses arrive on stdin.
type Client struct {
	write          FrameWriter
	mu             sync.Mutex
	pending        map[string]*pending
	disabledReason string
}

// New constructs a Client. Pass nil to use the default stdout writer.
func New(w FrameWriter) *Client {
	if w == nil {
		w = DefaultFrameWriter()
	}
	return &Client{
		write:   w,
		pending: map[string]*pending{},
	}
}

// Disable marks web access as unavailable.
func (c *Client) Disable(reason string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.disabledReason = reason
}

// Search issues a `web/search` request. Pass timeout = 0 for the default
// (30s).
func (c *Client) Search(req SearchRequest, timeout time.Duration) (*SearchResponse, error) {
	if req.Query == "" {
		return nil, errors.New("query must be non-empty")
	}
	if timeout <= 0 {
		timeout = 30 * time.Second
	}
	raw, err := c.call(MethodWebSearch, req, timeout)
	if err != nil {
		return nil, err
	}
	var out SearchResponse
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, fmt.Errorf("decode web/search result: %w", err)
	}
	return &out, nil
}

// Fetch issues a `web/fetch` request. Pass timeout = 0 for the default
// (90s — fetch may crawl up to 10 slow pages).
func (c *Client) Fetch(req FetchRequest, timeout time.Duration) (*FetchResponse, error) {
	if len(req.URLs) == 0 {
		return nil, errors.New("urls must be non-empty")
	}
	if timeout <= 0 {
		timeout = 90 * time.Second
	}
	raw, err := c.call(MethodWebFetch, req, timeout)
	if err != nil {
		return nil, err
	}
	var out FetchResponse
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, fmt.Errorf("decode web/fetch result: %w", err)
	}
	return &out, nil
}

// ImageSearch issues a `web/image_search` request (Phase 2). Requires
// manifest host_capabilities "web.image_search" + the user's
// web_grant.allowImageSearch (defaults OFF). Pass timeout = 0 for the
// default (30s).
func (c *Client) ImageSearch(req ImageSearchRequest, timeout time.Duration) (*ImageSearchResponse, error) {
	if req.Query == "" {
		return nil, errors.New("query must be non-empty")
	}
	if timeout <= 0 {
		timeout = 30 * time.Second
	}
	raw, err := c.call(MethodWebImageSearch, req, timeout)
	if err != nil {
		return nil, err
	}
	var out ImageSearchResponse
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, fmt.Errorf("decode web/image_search result: %w", err)
	}
	return &out, nil
}

// ImageFetch issues a `web/image_fetch` request (Phase 2): the host
// downloads the image (SSRF-guarded, MIME whitelist + magic bytes) into
// APS file storage and returns an artifact reference — never bytes.
// Requires "web.image_fetch" cap + web_grant.allowImageFetch (defaults
// OFF). Pass timeout = 0 for the default (60s).
func (c *Client) ImageFetch(req ImageFetchRequest, timeout time.Duration) (*ImageFetchResponse, error) {
	if req.URL == "" {
		return nil, errors.New("url must be non-empty")
	}
	if timeout <= 0 {
		timeout = 60 * time.Second
	}
	raw, err := c.call(MethodWebImageFetch, req, timeout)
	if err != nil {
		return nil, err
	}
	var out ImageFetchResponse
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, fmt.Errorf("decode web/image_fetch result: %w", err)
	}
	return &out, nil
}

func (c *Client) call(method string, params any, timeout time.Duration) (json.RawMessage, error) {
	c.mu.Lock()
	if c.disabledReason != "" {
		reason := c.disabledReason
		c.mu.Unlock()
		return nil, &WebError{Code: ErrCodeNotNegotiated, Message: reason}
	}
	c.mu.Unlock()

	id, err := newReqID()
	if err != nil {
		return nil, err
	}

	p := &pending{ch: make(chan json.RawMessage, 1)}
	c.mu.Lock()
	c.pending[id] = p
	c.mu.Unlock()

	envelope := map[string]any{
		"jsonrpc": "2.0",
		"id":      id,
		"method":  method,
		"params":  params,
	}
	if err := c.write(envelope); err != nil {
		c.mu.Lock()
		delete(c.pending, id)
		c.mu.Unlock()
		return nil, err
	}

	select {
	case raw := <-p.ch:
		var resp struct {
			Result json.RawMessage `json:"result"`
			Error  *WebError       `json:"error"`
		}
		if err := json.Unmarshal(raw, &resp); err != nil {
			return nil, fmt.Errorf("decode web response: %w", err)
		}
		if resp.Error != nil {
			return nil, resp.Error
		}
		if resp.Result == nil {
			return nil, errors.New("empty web result")
		}
		return resp.Result, nil
	case <-time.After(timeout):
		c.mu.Lock()
		delete(c.pending, id)
		c.mu.Unlock()
		return nil, &WebError{
			Code:    ErrCodeTimeout,
			Message: fmt.Sprintf("%s timed out after %s", method, timeout),
		}
	}
}

// DispatchResponse resolves a pending request from a parsed JSON-RPC
// frame. Returns true if `frame` was a response we owned.
func (c *Client) DispatchResponse(frame json.RawMessage) bool {
	var head struct {
		ID     any  `json:"id"`
		Method *any `json:"method"`
	}
	if err := json.Unmarshal(frame, &head); err != nil {
		return false
	}
	if head.Method != nil {
		return false
	}
	idStr, ok := head.ID.(string)
	if !ok {
		return false
	}
	c.mu.Lock()
	p := c.pending[idStr]
	if p != nil {
		delete(c.pending, idStr)
	}
	c.mu.Unlock()
	if p == nil {
		return false
	}
	select {
	case p.ch <- frame:
	default:
	}
	return true
}

func newReqID() (string, error) {
	buf := make([]byte, 16)
	if _, err := rand.Read(buf); err != nil {
		return "", err
	}
	return hex.EncodeToString(buf), nil
}

File: sdk/nodejs/web.js
/**
 * WebClient — issue reverse `web/search` and `web/fetch` JSON-RPC requests
 * to the host Agent (Anna), which proxies to Nexus's `/api/v1/copilot/web/*`
 * endpoints using the short-lived `sampling_token` the host minted at
 * invoke time.
 *
 * The plugin never sees a search-provider API key — provider routing,
 * SSRF guarding, quota, billing and audit all stay host-side. Response
 * schemas are provider-agnostic (`search_depth` expresses quality intent,
 * never provider choice).
 *
 * stdio-channel budgets (tighter than the iframe HOST API):
 *   - fetch: max_chars ≤ 8,000/page, whole response ≤ 256 KB.
 *
 * Threading model identical to SamplingClient / ImageClient:
 *   - Construct one WebClient per process.
 *   - Feed every parsed JSON-RPC frame received on stdin to
 *     `web.dispatchResponse(msg)`; pair with `makeResponseRouter()`.
 *
 * Error codes — keep in sync with matrix/src/executa/protocol.py:
 *   WEB_ERR_NOT_GRANTED     = -32521
 *   WEB_ERR_QUOTA_EXCEEDED  = -32522   (CU pool exhausted)
 *   WEB_ERR_PROVIDER_ERROR  = -32523
 *   WEB_ERR_INVALID_REQUEST = -32524
 *   WEB_ERR_TIMEOUT         = -32525
 *   WEB_ERR_NOT_NEGOTIATED  = -32526
 *   WEB_ERR_USER_DENIED     = -32527
 *   WEB_ERR_TOKEN_EXPIRED   = -32528
 *
 * See matrix-nexus docs/design/app-web-search.md §2 / §4 for the wire
 * contract.
 */

"use strict";

const crypto = require("node:crypto");

const { attachInvokeContext } = require("./context");

const METHOD_WEB_SEARCH = "web/search";
const METHOD_WEB_FETCH = "web/fetch";
// Phase 2 (app-web-search.md §2.3 / §2.4)
const METHOD_WEB_IMAGE_SEARCH = "web/image_search";
const METHOD_WEB_IMAGE_FETCH = "web/image_fetch";

const WEB_ERR_NOT_GRANTED = -32521;
const WEB_ERR_QUOTA_EXCEEDED = -32522;
const WEB_ERR_PROVIDER_ERROR = -32523;
const WEB_ERR_INVALID_REQUEST = -32524;
const WEB_ERR_TIMEOUT = -32525;
const WEB_ERR_NOT_NEGOTIATED = -32526;
const WEB_ERR_USER_DENIED = -32527;
const WEB_ERR_TOKEN_EXPIRED = -32528;

class WebError extends Error {
  constructor(code, message, data) {
    super(`[${code}] ${message}`);
    this.name = "WebError";
    this.code = code;
    this.data = data || {};
  }
}

class WebClient {
  constructor(opts = {}) {
    this._writeFrame =
      opts.writeFrame ||
      ((msg) => {
        process.stdout.write(JSON.stringify(msg) + "\n");
      });
    /** @type {Map<string, {resolve: Function, reject: Function, timer: NodeJS.Timeout}>} */
    this._pending = new Map();
    this._disabledReason = null;
  }

  disable(reason) {
    this._disabledReason = reason;
  }

  /**
   * Web / news search. Requires manifest
   * `host_capabilities: ["web.search"]` plus the user's web grant.
   *
   * Resolves to `{ results: [{title, url, snippet, site, published_at?,
   * score?}], provider_tier: "basic"|"advanced", quota_consumed }`.
   * `provider_tier` reports the ACTUALLY executed tier (a degraded
   * advanced request comes back "basic" and is billed cheaper).
   *
   * @param {{
   *   query: string,
   *   max_results?: number,              // default 5
   *   search_depth?: "basic"|"advanced", // advanced needs allowAdvanced grant
   *   topic?: "general"|"news",
   *   time_range?: "day"|"week"|"month"|"year",
   *   region?: string,                   // e.g. "us-en"
   *   include_domains?: string[],
   *   exclude_domains?: string[],
   *   timeoutMs?: number,                // default 30_000
   * }} opts
   */
  search(opts) {
    const {
      query,
      max_results,
      search_depth,
      topic,
      time_range,
      region,
      include_domains,
      exclude_domains,
      timeoutMs = 30_000,
    } = opts;
    if (typeof query !== "string" || !query.trim()) {
      return Promise.reject(
        new WebError(WEB_ERR_INVALID_REQUEST, "query must be non-empty string")
      );
    }
    const params = { query };
    if (max_results != null) params.max_results = Number(max_results);
    if (search_depth != null) params.search_depth = search_depth;
    if (topic != null) params.topic = topic;
    if (time_range != null) params.time_range = time_range;
    if (region != null) params.region = region;
    if (include_domains != null) params.include_domains = include_domains;
    if (exclude_domains != null) params.exclude_domains = exclude_domains;
    return this._call(METHOD_WEB_SEARCH, params, timeoutMs);
  }

  /**
   * SSRF-guarded page extraction (1–10 URLs, per-item failure isolation).
   * Requires manifest `host_capabilities: ["web.fetch"]` + user grant.
   *
   * Resolves to `{ pages: [{url, final_url?, ok, title?, content?,
   * truncated?, error?}], quota_consumed }` — check `pages[i].ok` per
   * item; the whole call only rejects on auth/quota/protocol errors.
   *
   * @param {{
   *   urls: string[],
   *   format?: "markdown"|"text",
   *   max_chars?: number,      // stdio cap: 8,000 per page
   *   timeout_ms?: number,     // per-page budget, ≤ 30_000
   *   timeoutMs?: number,      // client wall clock, default 90_000
   * }} opts
   */
  fetch(opts) {
    const { urls, format, max_chars, timeout_ms, timeoutMs = 90_000 } = opts;
    if (!Array.isArray(urls) || urls.length === 0) {
      return Promise.reject(
        new WebError(WEB_ERR_INVALID_REQUEST, "urls must be non-empty array")
      );
    }
    const params = { urls };
    if (format != null) params.format = format;
    if (max_chars != null) params.max_chars = Number(max_chars);
    if (timeout_ms != null) params.timeout_ms = Number(timeout_ms);
    return this._call(METHOD_WEB_FETCH, params, timeoutMs);
  }

  /**
   * Image search (Phase 2). Requires manifest
   * `host_capabilities: ["web.image_search"]` plus the user's
   * `web_grant.allowImageSearch` (defaults OFF). Safe-search is always
   * enforced host-side and cannot be disabled. `license_hint` is a
   * best-effort passthrough — the platform makes no copyright warranty.
   *
   * Resolves to `{ results: [{image_url, thumbnail_url?, source_url,
   * title?, width?, height?, mime_type?, license_hint?}],
   * quota_consumed, cached? }`.
   *
   * @param {{
   *   query: string,
   *   max_results?: number,            // default 8, cap 20
   *   min_width?: number, min_height?: number,
   *   aspect?: "any"|"wide"|"tall"|"square",
   *   timeoutMs?: number,              // default 30_000
   * }} opts
   */
  imageSearch(opts) {
    const { query, max_results, min_width, min_height, aspect, timeoutMs = 30_000 } = opts;
    if (typeof query !== "string" || !query.trim()) {
      return Promise.reject(
        new WebError(WEB_ERR_INVALID_REQUEST, "query must be non-empty string")
      );
    }
    const params = { query };
    if (max_results != null) params.max_results = Number(max_results);
    if (min_width != null) params.min_width = Number(min_width);
    if (min_height != null) params.min_height = Number(min_height);
    if (aspect != null) params.aspect = aspect;
    return this._call(METHOD_WEB_IMAGE_SEARCH, params, timeoutMs);
  }

  /**
   * Image download → APS files artifact (Phase 2). Requires manifest
   * `host_capabilities: ["web.image_fetch"]` + `web_grant.allowImageFetch`
   * (defaults OFF). The response is an artifact REFERENCE (`path` +
   * short-lived `get_url`), never bytes — stdio-frame safe. The stored
   * object counts against the tool's storage quota.
   *
   * Resolves to `{ path, get_url, mime_type, bytes_size, sha256,
   * source_url, final_url, quota_consumed }`.
   *
   * @param {{
   *   url: string,
   *   max_bytes?: number,   // default 5 MiB, cap 20 MiB
   *   purpose?: string,     // audit label
   *   timeoutMs?: number,   // default 60_000
   * }} opts
   */
  imageFetch(opts) {
    const { url, max_bytes, purpose, timeoutMs = 60_000 } = opts;
    if (typeof url !== "string" || !/^https?:\/\//.test(url)) {
      return Promise.reject(
        new WebError(WEB_ERR_INVALID_REQUEST, "url must be an http(s) URL")
      );
    }
    const params = { url };
    if (max_bytes != null) params.max_bytes = Number(max_bytes);
    if (purpose != null) params.purpose = purpose;
    return this._call(METHOD_WEB_IMAGE_FETCH, params, timeoutMs);
  }

  dispatchResponse(msg) {
    if (!msg || typeof msg !== "object" || "method" in msg) return false;
    const id = msg.id;
    if (id == null) return false;
    const pending = this._pending.get(id);
    if (!pending) return false;
    this._pending.delete(id);
    clearTimeout(pending.timer);
    if (msg.error) {
      pending.reject(
        new WebError(
          Number(msg.error.code) || -32603,
          String(msg.error.message || "unknown error"),
          msg.error.data
        )
      );
    } else {
      pending.resolve(msg.result || {});
    }
    return true;
  }

  _call(method, params, timeoutMs) {
    if (this._disabledReason) {
      return Promise.reject(
        new WebError(WEB_ERR_NOT_NEGOTIATED, this._disabledReason)
      );
    }
    const reqId = crypto.randomUUID();
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        if (this._pending.delete(reqId)) {
          reject(
            new WebError(
              WEB_ERR_TIMEOUT,
              `${method} timed out after ${timeoutMs}ms`
            )
          );
        }
      }, timeoutMs);
      this._pending.set(reqId, { resolve, reject, timer });
      try {
        this._writeFrame({
          jsonrpc: "2.0",
          id: reqId,
          method,
          params: attachInvokeContext(params),
        });
      } catch (err) {
        clearTimeout(timer);
        this._pending.delete(reqId);
        reject(err);
      }
    });
  }
}

module.exports = {
  WebClient,
  WebError,
  METHOD_WEB_SEARCH,
  METHOD_WEB_FETCH,
  METHOD_WEB_IMAGE_SEARCH,
  METHOD_WEB_IMAGE_FETCH,
  WEB_ERR_NOT_GRANTED,
  WEB_ERR_QUOTA_EXCEEDED,
  WEB_ERR_PROVIDER_ERROR,
  WEB_ERR_INVALID_REQUEST,
  WEB_ERR_TIMEOUT,
  WEB_ERR_NOT_NEGOTIATED,
  WEB_ERR_USER_DENIED,
  WEB_ERR_TOKEN_EXPIRED,
};
